cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java

```diff
@@ -917,4 +931,108 @@ private boolean deleteIndexRow(Row row) {
+    }
+
+    return entries;
+  }
+
+  private List<KeyValue<Long, MetadataEntry>> scanOrDelete(byte[] rowPrefix, byte[] stopRowKey,
+                                                           int limit, boolean delete,
+                                                           Function<Row, KeyValue<Long, MetadataEntry>> function) {
+    List<KeyValue<Long, MetadataEntry>> entries = new LinkedList<>();
+    try (Scanner scan = indexedTable.scan(rowPrefix, stopRowKey)) {
+      Row row;
+      while ((row = scan.next()) != null && limit > 0) {
+        KeyValue<Long, MetadataEntry> entry = function.apply(row);
+
+        if (entry != null) {
+          entries.add(entry);
+          limit--;
+        }
+
+        if (delete) {
+          indexedTable.delete(new Delete(row.getRow()));
+        }
+      }
+    }
+    return entries;
+  }
+
+  /**
+   * Deserializes value row key to get metadata entry.
+   *
+   * @param row value row.
+   * @return {@link KeyValue} key is null, value is {@link MetadataEntry}.
+   */
+  @Nullable
+  private KeyValue<Long, MetadataEntry> convertFromV1Value(Row row) {
+    byte[] rowKey = row.getRow();
+    String targetType = MdsKey.getTargetType(rowKey);
+    NamespacedEntityId namespacedEntityId = MdsKey.getNamespacedIdFromKey(targetType, rowKey);
+    String key = MdsKey.getMetadataKey(targetType, rowKey);
+    byte[] value = row.get(VALUE_COLUMN);
+    if (key == null || value == null) {
+      return null;
+    }
+    return new KeyValue<>(null, new MetadataEntry(namespacedEntityId, key, Bytes.toString(value)));
+  }
+
+  /**
+   * Deserializes history row key to get timestamp and metadata entry.
+   *
+   * @param row history row.
+   * @return {@link KeyValue} key is timestamp from history key, value is {@link MetadataEntry}.
+   */
+  @Nullable
+  private KeyValue<Long, MetadataEntry> convertFromV1History(Row row) {
+    byte[] rowKey = row.getRow();
+    long historyTime = MdsHistoryKey.getHistoryTime(rowKey);
+    MetadataV1 metadata = GSON.fromJson(row.getString(HISTORY_COLUMN), MetadataV1.class);
+    if (metadata == null) {
+      return null;
+    }
+    // For history we do not care about key and value for MetadataEntry as we are not going to extract that
+    // information for upgrade
+    return new KeyValue<>(historyTime, new MetadataEntry(metadata.getEntityId(), "", ""));
+  }
+
+  /**
+   * Writes entries to V2 MetadataTable.
+   *
+   * @param entries list of entries to be written.
+   */
+  public void writeUpgradedRow(List<KeyValue<Long, MetadataEntry>> entries) {
+    for (KeyValue<Long, MetadataEntry> kv : entries) {
+      MetadataEntry entry = kv.getValue();
+      if (kv.getKey() == null) {
+        writeWithoutHistory(entry.getMetadataEntity(), entry,
+                            getIndexersForKey(entry.getMetadataEntity(), entry.getKey()));
+      } else {
+        writeHistory(entry.getMetadataEntity());
```

Comment:
i think you want to call ```writeHistory(entry.getMetadataEntity(), kv.getKey()));``` so that the upgraded history entry has the same time as the original one?